### DISCUSSION TITLE
Ideal workflow setup stage

### DISCUSSION THREAD

**Discussion by atherdon** [atherdon/dash-back#51] [General]
Soon we'll need to have a workflow that will work not only on local machines

It should be simple:
1) each time when we merge PR for backend - it should be automatically deployed "to the cloud"
2) upload forms project should be also deployed and sync with the latest DB schema changes. Also should be publicly available online, so Marina can easily open it and save the information into our deployed database
3) data, that was uploaded via upload form, should be viewable in our dashboard-react. 
4) dashboard-react should trigger errors, when the updated DB schema didn't match with stuff that we have inside of our frontend.

**Comment by kolserdav** [comment]
>dashboard-react should trigger errors, when the updated DB schema didn't match with stuff that we have inside of our frontend.
>
yes a error have where the types do not match, but in `data` some data, for example `published`, met both as a string and as a number, so they are made strings

**Comment by atherdon** [comment]
@kolserdav вот этот вопрос недает мне покоя. если бы мы могли настроить "идеальный" coding pipeline, aka DevOps/CI-CD мне бы сильно это упростило жизнь. Я не уверен что с этой проблемой я должен идти к вам - мне кажется что нам нужен отдельный "админ" который должен это настроить, но может у вас есть идеи как этого добится.

**Reply by kolserdav** [reply]
Надо сначала сервер подготовить чтобы там работало, а потом гуглить как с git синхронизировать, у меня руки не доходили ещё до этого, вручную всегда деплой делал. Теоретически могу предположить, что скрипт через крон каждую минуту проверяет по git api версию и в случае обновы пересобирает.

**Comment by atherdon** [comment]
Скорее всего - наша п проблема является тривиальной, и уже есть решения,
которые позволяют делать всякие sync между проектами и ТД. Я все ещё смотрю
в сторону aws и их тулзовин, которые работают с devops. Мы конечно можем
разбить проблему на части. Может нам просто настроить логику деплоя на aws
по одной команде

чт, 4 мар. 2021 г., 4:33 AM kolserdav <notifications@github.com>:

**Reply by kolserdav** [reply]
Я одно время штурмовал aws но потом сломалась двухфакторная аутентификация, а на номер, который там указан они не могут дозвониться хотя он работает)

**Comment by atherdon** [comment]
https://medium.com/innovation-incubator/deploying-code-from-github-to-aws-ec2-with-codepipeline-4639e8fbba28

При этом у призмы другой пример как работать с их фреймворком
https://www.prisma.io/docs/guides/deployment/deploying-to-aws-lambda


чт, 4 мар. 2021 г., 2:26 PM Arthur Tkachenko <
arthur.tkachenko.netweight@gmail.com>:

**Reply by kolserdav** [reply]
Это разные случаи. Один EC2 второй Lambda.
У меня с Amazon не складывается пока что(
Поэтому скажу свое мнение:
Но я особо не переживаю, так как тогда сделал для себя вывод, что на Lambda особо не съэкономишь, так как стоимость разработки на нем повышается из за специфических особенностей настройки окружения, плюс там желательно сразу брать платную поддержку, иначе руки гуглить устанут, если что-то пойдет не так. Lambda на мой взгляд имеет одно весомое преимущество, благодаря которому может конкурировать, на некоторых проектах, с сервером под рукой - это его автоматическая масштабируемость. Когда сегодня 3 человека зашли а завтра 30000, он не упадет в отличие от сервера, рассчитанного "с запасом" на 3000. А EC2 обычный VPS, далеко не самый доступный по цене, в разы выше чем чем на более народных сервисах, из преимуществ я заметил только большой выбор образов и практически неограниченная масштабируемость. Но суммы меня пугают, нужно наверное жить в Los-Angeles, чтобы привыкнуть к этому. Поэтому, я остался не в восторге и фраза "То что можно сделать бесплатно, продается там" у меня никогда не выходит из головы, когда про них думаю. Но это только моё мнение)

**Reply by atherdon** [reply]
Я поэтому тоже незнаю что выбрать.
но наверное мы будем делать вначале быстрый вариант а потом уже думать о нормальном варианте. Я пытался найти DevOps на FH - пока у меня прогресса ноль. я думал что они как девы и я смогу найти с ними общий язык, но пока что ничего не получается. значит будем делать по старинке. вначале мелкие проблемы будем решать, а потом уже придумаем нормальное worfklow.

я просто ночью осознал что я с 10го марта иду в отпуск. и надо одну из версий dashboard сделать рабочей. чтобы в ней могла работать и Марина и другие редакторы. поэтому наверное все идеи о хорошем workflow переносятся на то время, когда я уже вернусь - может ктото мне и напишет сам на FH за это время.

поэтому я думаю что мы выберем пока heroku - а там будет видно.

**Reply by atherdon** [reply]
Я зарегаю heroku и скину вам доступы. они норм типы - еще и спонсировали нашу рассылку год назад.
так будет быстрее. поидее мы сможем повисеть какое то время на дешевом плане, а дальше разберемся

**Reply by kolserdav** [reply]
врядли на heroku получится PrismaStudio запустить

**Reply by atherdon** [reply]
создаю отдельный таск под heroku. чтобы проще было

**Reply by atherdon** [reply]
https://github.com/atherdon/dash-back/issues/89

**Comment by kolserdav** [comment]
Может вариант с Docker подойдет. просто чтобы на сервере был докер и одной коммандой все разворачивалось, раз уж есть проблема с настройкой окружения на сервере?

**Reply by atherdon** [reply]
думал об этом тоже. как сложно это будет сделать?

**Reply by kolserdav** [reply]
Настроить Dockerfile для `dash-back` , `dash-upload-form`, `nginx` - если нужен ssl. подключить контейнер mysql, переделать на mysql призму. Один раз надо будет создать ssl сертификат например с letencrypt и `nginx.conf` настроить (если нужен ssl), и контейнер `phpMyAdmin` для наглядности данных. Часа 3 уйдет на настройку и тестирование всего этого.

**Reply by kolserdav** [reply]
У нас и проект уже mysql просит, на sqlite не поддерживается тип enum, он нужен для типов статей

**Reply by kolserdav** [reply]
> 
> 
> Настроить Dockerfile для `dash-back` , `dash-upload-form`, `nginx` - если нужен ssl. подключить контейнер mysql, переделать на mysql призму. Один раз надо будет создать ssl сертификат например с letencrypt и `nginx.conf` настроить (если нужен ssl), и контейнер `phpMyAdmin` для наглядности данных. Часа 3 уйдет на настройку и тестирование всего этого.

то есть `docker-compose.yml` и на сервере Docker и docker-compose

**Reply by atherdon** [reply]
@kolserdav mysql нехотел юзать. оно мне напоминает о php :) я думал про postgre.

**Reply by kolserdav** [reply]
Можно postgre и наверное pgadmin
